src/libraries/System.Private.CoreLib/src/System/Double.cs

```diff
@@ -1186,13 +1164,462 @@ public static double MinMagnitudeNumber(double x, double y)
             return y;
         }
 
-        /// <inheritdoc cref="INumberBase{TSelf}.TryCreate{TOther}(TOther, out TSelf)" />
+        /// <inheritdoc cref="INumberBase{TSelf}.TryConvertFromChecked{TOther}(TOther, out TSelf)" />
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static bool TryCreate<TOther>(TOther value, out double result)
-            where TOther : INumberBase<TOther>
+        static bool INumberBase<double>.TryConvertFromChecked<TOther>(TOther value, out double result)
+        {
+            // In order to reduce overall code duplication and improve the inlinabilty of these
+            // methods for the corelib types we have `ConvertFrom` handle the same sign and
+            // `ConvertTo` handle the opposite sign. However, since there is an uneven split
+            // between signed and unsigned types, the the one that handles unsigned will also
+            // handle `Decimal` and `NFloat`.
```

Comment:
Would it be better for linking to have all Decimal and NFloat specific conversions in Decimal and NFloat, so that Decimal or NFloat can be trimmed in apps that do not otherwise use them?

Reply:
(Same concern for UInt128/Int128.)

Reply:
We have 18 numeric types in corelib and they need to convert to/from each other with a given `TSelf.Create*` needing to call both `TSelf.TryConvertFrom*` and `TOther.TryConvertTo*`. Because of this there isn't really any setup where using one of the types doesn't "keep" the other 17 types if any `Create` call gets used.

Likewise we want to try and keep the method bodies "small" overall to help with inlining and JIT throughput when generic math gets used. Having `decimal` or `NFloat` convert all 17 types in `ConvertTo` will hinder their scenarios as the IL size, even with the inlining improvements, appears to be "massive" (even though realistically only one path will ever be kept).

I think it'd be better for the trimmer to be smarter around the `if (typeof(T) == typeof(...))` pattern when that's the only usage and/or for the JIT to get improvements here first so we know its not going to be problematic.

Reply:
I agree that the split between ConvertFrom/ConvertTo is reasonable to core primitive types that are always used.

The non-core types like `NFloat` that are often not used by the app should be on the same plan as e.g. `Complex`.

Reply:
Where does that leave `decimal` (C# primitive and only currency like type), `Half` (Machine Learning), `Int128`, and `UInt128 (ABI primitives that have long been requested)?

Reply:
> where using one of the types doesn't "keep" the other 17 types if any Create call gets used.

How does `Complex` avoid the problem with everything depending on everything else?

Reply:
`Complex` and `BigInteger` have very big method bodies because they sit at a higher layer and so can't do this "splitting" trick.

At best we could factor it out into a couple smaller helper methods (e.g. `if (value is IBinaryInteger) { TryConvertFromBinaryIntegerChecked(...) } else if (value is IFloatingPointe) { .... }`) to try and improve what works with the existing inlining support.

Otherwise we'll likely need some improvement in the JIT. That was "out of scope" for this PR which was targeting P5 (but has missed the snap now of course)

Reply:
I think we should reduce the bubble of 17 types where everything depends on everything else to make this more linker friendly.

I am sure we can find some reasonable way to workaround the JIT inliner limitations.

cc @marek-safar

Reply:
Once you do that, NFloat can move back to S.R.InteropServices where it logically belongs.

Reply:
Would `System.Runtime.Numerics` then depend on `System.Runtime.InteropServices` or are you suggesting that you can only convert `NFloat` to/from `BigInteger`/`Complex` using an external converter?

Reply:
Either way would be fine with me. If I were to choose, I would omit the built-in conversion for `NFloat` to/from BigInteger/Complex.

Reply:
Moved it back to S.R.InteropServices and adjusted the API surface and implementation accordingly

Reply:
Can we start checking the size impact for changes like this?

Reply:
Are there instructions for how to do that?

Reply:
It would be possible, but I think it would be better for the linker to recognize and get special handling for the `if (typeof(T) == typeof(...))` pattern as it is a broader and more widely applicable thing for these types of issues.